doc/source/development/rfc/rfc100_float16_support.rst

```diff
@@ -0,0 +1,141 @@
+============== =============================================
+
+Summary
+-------
+
+This RFC adds support for the IEEE 16-bit floating point data type
+(aka ``half``, ``float16``). It adds new pixel data types
+``GDT_Float16`` and ``GDT_CFloat16``, backed by a new ``GFloat16`` C++
+type. This type will be emulated in software if not available
+natively.
+
+Some drivers, in particular Zarr, will be extended to support this
+type. Drivers that do not support this type will be checked to ensure
+they report an appropriate errors when necessary.
+
+Motivation
+----------
+
+- The float16 type is increasingly common. It is supported by GPUs, by
+  modern CPUs, and by modern compilers (e.g. GCC 12).
+
+- Using ``uint16`` values to shuffle ``float16`` values around is
+  inconvenient and awkward. There is no mechanism in GDAL to mark
+  attributes or datasets which ``uint16`` values should be interpreted
+  as ``float16``.
+
+- Some drivers (at least the HDF5, GTiff, and Zarr) already handle
+  float16 by exposing it as float32, using software conversion
+  routines. This type should be supported without converting to
+  float32.
+
+- C++23 will introduce native support for ``std::float16_t``. However,
+  it will likely be several years until C++23 will be a requirement
+  for GDAL. A shorter-term solution is needed.
+
+- Many other projects and storage libraries support float16. GDAL will
+  fall behind if it doesn't.
+
+Details
+-------
+
+A new type ``GFloat16`` will be defined in C++. This type will be a
+thin wrapper around a native float16 type if one is available, such as
+e.g. ``_Float16`` in modern versions of GCC. Otherwise, a this type
+will be emulated (transparently to the user), and operations will be
+performed as ``float``.
+
+The following pixel data types are added:
+- ``GDT_Float16``  --> ``_Float16``
+- ``GDT_CFloat16`` --> ``std::complex<_Float16``
+
+Some drivers (at least the HDF5, GTiff, and Zarr) already handle
+float16 by exposing it as float32, using software conversion routines.
+float16 is now supported directly, i.e., without converting to
+float32. In other drivers, the current behaviour is retained, which
+automatically converts float16 to float32.
+
+For simplicity there will be no new functions handling attributes for
+multidimensional datasets. Attributes of type float16 can still be
+read/written as raw values.
+
+Impacts in the code base
+------------------------
+
+It is likely that a large fraction of the code base will be impacted.
+
+SWIG bindings
+-------------
+
+The SWIG bindings are extended where possible. Unfortunately, SWIG
+does not support the native float16 Python type, but it does support
+the float16 numpy type. This means that not all SWIG Python wrappers
+can support float16.
+
+Backward compatibility
+----------------------
+
+C and C++ API and ABI are impacted. This design is backward-compatible
+manner, i.e. it does not break the ABI, and can thus be implemented in
+GDAL 3.x.
```

Comment:
We'll need a section though in the MIGRATION_GUIDE.TXT similar to when we introduced Int8 and Int64/UInt64, because code that use the GDAL API must be ready to react to the new data type, possibly by doing RasterIO() requests with eBufType==GDT_Float32, if they can't deal natively with Float16 values